README.md
# text_3d: 3D Text Effects for Flutter

Easily incorporate visually captivating 3D text effects into your Flutter applications.
The `text_3d` library offers a range of predefined styles, and the flexibility to customize, letting
your creativity roam free.

![Flutter](https://img.shields.io/badge/Flutter%20-%2302569B.svg?style=flat&logo=flutter&logoColor=FFFFFF)
![Dart](https://img.shields.io/badge/dart-%230175C2.svg?style=flat&logo=dart&logoColor=FFFFFF)
## Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [Available Styles](#available-styles)
- [Example](#example)
- [Custom Styles](#custom-styles)
- [Contributing](#contributing)
- [License](#license)

## Installation

Add the following line to your `pubspec.yaml` under `dependencies`:

```
text_3d: latest_version
```

Then run:

```
flutter pub get
```

## Usage

To use the `ThreeDText` widget:

```dart
import 'package:text_3d/text_3d.dart';
```

## Available Styles

- `standard`: A basic 3D effect.
- `raised`: Text appears raised from the background.
- `inset`: Text appears pressed into the background.
- `perspectiveRaised`: Text has a raised perspective look.
- `perspectiveInset`: Text has a pressed perspective look.
- `perspectiveLeft`: Text slants to the left in perspective.
- `perspectiveRight`: Text slants to the right in perspective.

## Example

Here's a quick example of how to use the `ThreeDText` widget:

![Example Image](https://github.com/setayeshce/text3d/blob/main/assets/images/perspectiveRaised.png?raw=true)


```dart
 ThreeDText(
text: 'perspectiveRaised',
textStyle: TextStyle(fontSize: 40, color: Colors.green),
depth: 10,
style: ThreeDStyle.perspectiveRaised,
angle: pi / 6,
perspectiveDepth: 30, 
),

```

![Example Image](https://github.com/setayeshce/text3d/blob/main/assets/images/perspectiveInset.png?raw=true)


```dart
 ThreeDText(
text: 'perspectiveInset',
textStyle: TextStyle(fontSize: 40, color: Colors.purpleAccent),
depth: 40,
style: ThreeDStyle.perspectiveInset,
angle: pi / 6,
perspectiveDepth: 20,
),
```

![Example Image](https://github.com/setayeshce/text3d/blob/main/assets/images/perspectiveRight.png?raw=true)

```dart
ThreeDText(
text: 'perspectiveRight',
textStyle: TextStyle(fontSize: 25, color: Colors.yellow),
style: ThreeDStyle.perspectiveLeft,
perspectiveDepth: 45.0,
),
```

![Example Image](https://github.com/setayeshce/text3d/blob/main/assets/images/perspectiveLeft.png?raw=true)


```dart
  ThreeDText(
text: 'perspectiveLeft',
textStyle: const TextStyle(
fontSize: 25,
color: Colors.pink ,
fontWeight: FontWeight.bold),
depth: 6,
style: ThreeDStyle.perspectiveLeft,
perspectiveDepth: -45.0
),
```

![Example Image](https://github.com/setayeshce/text3d/blob/main/assets/images/inset.png?raw=true)

```dart
ThreeDText(
text: "inset",
textStyle: TextStyle(
fontSize: 64,
color: Colors.pinkAccent,
),
style: ThreeDStyle.inset,
),
```


![Example Image](https://github.com/setayeshce/text3d/blob/main/assets/images/raised.png?raw=true)

```dart
ThreeDText(
text: 'raised',
textStyle: TextStyle(fontSize: 64, color: Colors.blue),
depth: 10,
style: ThreeDStyle.raised,
),
```

![Example Image](https://github.com/setayeshce/text3d/blob/main/assets/images/standard.png?raw=true)


```dart
ThreeDText(
text: 'standard',
textStyle: TextStyle(fontSize: 40, color: Colors.blue),
depth: 5,
style: ThreeDStyle.standard,
),
```

![Example Image](https://github.com/setayeshce/text3d/blob/main/assets/images/insetAngle.png?raw=true)


```dart
ThreeDText(
text: 'inset with angle',
textStyle: TextStyle(fontSize: 40, color: Colors.deepOrange),
depth: 5,
style: ThreeDStyle.inset,
angle: 70,
),
```

## Custom Styles

While the library comes packed with a range of predefined styles, it's designed with flexibility in
mind. You can adjust properties like `angle`, `depth`, and `perspectiveDepth` to create unique 3D
effects tailored to your design vision.

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would
like to change.

## License

[MIT](https://choosealicense.com/licenses/mit/)

---

Copy the above text into your README.md file, and it should be well-documented and formatted for
users and developers alike!

CHANGELOG.md
## [0.0.4] - YYYY-MM-DD
### Added
- New features or changes in this version.

* TODO: Describe initial release.


test/tex_3d_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:text_3d/text_3d.dart';

void main() {
  // This group is for basic instantiation tests
  group('ThreeDText instantiation', () {
    testWidgets('should instantiate with standard style without errors',
        (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: ThreeDText(
          text: 'Test',
          textStyle: TextStyle(fontSize: 24, color: Colors.blue),
          style: ThreeDStyle.standard,
        ),
      ));
    });

    testWidgets('should instantiate with raised style without errors',
        (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: ThreeDText(
          text: 'Test',
          textStyle: TextStyle(fontSize: 24, color: Colors.blue),
          style: ThreeDStyle.raised,
        ),
      ));
    });

    // Add similar tests for the other styles...
  });

  // This group can be used to test other properties or behaviors
  group('ThreeDText properties and behaviors', () {
    testWidgets('should handle custom depth without errors',
        (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: ThreeDText(
          text: 'Test Depth',
          textStyle: TextStyle(fontSize: 24, color: Colors.red),
          depth: 15.0,
          style: ThreeDStyle.inset,
        ),
      ));
    });
  });
}


lib/text_3d.dart
// Library declaration for the 3D text effect.
library text_3d;

import 'dart:ui';
import 'package:flutter/material.dart';
import 'dart:math';

// Define the different styles available for 3D text.
enum ThreeDStyle {
  standard,
  raised,
  inset,
  perspectiveRaised,
  perspectiveInset,
  perspectiveLeft,
  perspectiveRight,
}

// Main widget for rendering 3D text.
class ThreeDText extends StatelessWidget {
  final String text;
  final TextStyle textStyle;
  final double depth; // Depth defines the 'thickness' of the 3D effect.
  final ThreeDStyle style; // The style of 3D effect.
  final double angle; // Angle of shadow layers
  final double perspectiveDepth; // Perspective depth for some effects

  ThreeDText({
    required this.text,
    required this.textStyle,
    this.depth = 10.0,
    this.style = ThreeDStyle.standard,
    this.angle = 0.0,
    this.perspectiveDepth = 300.0, // Default perspective depth
  });

  @override
  Widget build(BuildContext context) {
    List<Widget> layers = [];

    // Choose the appropriate 3D effect based on the style.
    switch (style) {
      case ThreeDStyle.standard:
        layers.addAll(_standardLayers());
        break;
      case ThreeDStyle.raised:
        layers.addAll(_raisedLayers());
        break;
      case ThreeDStyle.inset:
        layers.addAll(_insetLayers());
        break;
      default:
        layers.addAll(_perspectiveLayers());
    }

    return Stack(children: layers);
  }

  // Function to paint the 3D effect on the canvas.
  void paint(Canvas canvas, Size size) {
    final textSpan = TextSpan(text: text, style: textStyle);
    final textPainter =
        TextPainter(text: textSpan, textDirection: TextDirection.ltr);

    for (int i = 1; i <= depth; i++) {
      textPainter.paint(canvas, Offset(i * sin(angle), i * cos(angle)));
    }
    textPainter.paint(canvas, Offset(0, 0)); // Draw the main text
  }

  // Calculate offset based on the angle for 3D shadow.
  Offset _calculateOffset(int i) {
    return Offset(i * sin(angle), i * cos(angle));
  }

  // Normalize the layer number to get 3D effect.
  Offset _calculate3DOffset(int layer) {
    double factor = layer / depth;
    return Offset(lerpDouble(0, _calculateOffset(layer).dx, factor)!,
        lerpDouble(0, _calculateOffset(layer).dy, factor)!);
  }

  // Generate perspective transformation matrix.
  Matrix4 _perspectiveMatrix() {
    return Matrix4.identity()..setEntry(3, 2, 1 / perspectiveDepth);
  }

  // Layers for the standard 3D effect.
  List<Widget> _standardLayers() {
    return [
      for (int i = 1; i <= depth; i++)
        Transform.translate(
          offset: _calculate3DOffset(i),
          child: Text(
            text,
            style: textStyle.copyWith(color: Colors.black.withOpacity(0.2)),
          ),
        ),
      Text(text, style: textStyle),
    ];
  }

  // Layers for the raised 3D effect.
  List<Widget> _raisedLayers() {
    return [
      Text(text, style: textStyle),
      for (int i = 1; i <= depth; i++)
        Transform.translate(
          offset: _calculate3DOffset(i),
          child: Text(
            text,
            style: textStyle.copyWith(color: Colors.black.withOpacity(0.1)),
          ),
        ),
    ];
  }

  // Layers for the inset 3D effect.
  List<Widget> _insetLayers() {
    return [
      for (int i = 1; i <= depth; i++)
        Transform.translate(
          offset: _calculate3DOffset(-i),
          child: Text(
            text,
            style: textStyle.copyWith(color: Colors.black.withOpacity(0.2)),
          ),
        ),
      Text(text, style: textStyle),
    ];
  }

  // Layers for various perspective effects.
  List<Widget> _perspectiveLayers() {
    var transformation = _perspectiveMatrix();

    // Adjust transformation based on the style.
    switch (style) {
      case ThreeDStyle.perspectiveRaised:
        transformation.rotateX(0.2);
        break;
      case ThreeDStyle.perspectiveInset:
        transformation.rotateX(-0.2);
        break;
      case ThreeDStyle.perspectiveLeft:
        transformation.rotateY(0.2);
        break;
      case ThreeDStyle.perspectiveRight:
        transformation.rotateY(-0.2);
        break;
      default:
        break;
    }

    return [
      for (int i = 1; i <= depth; i++)
        Transform(
          alignment: FractionalOffset.center,
          transform: transformation,
          child: Transform.translate(
            offset: _calculate3DOffset(i),
            child: Text(
              text,
              style: textStyle.copyWith(
                  color: Colors.black.withOpacity(
                      0.1 / i)), // Fading the shadow makes it more realistic
            ),
          ),
        ),
      Transform(
        alignment: FractionalOffset.center,
        transform: transformation,
        child: Text(text, style: textStyle),
      ),
    ];
  }
}


example/lib/main.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:text_3d/text_3d.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
          child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          ThreeDText(
            text: 'perspectiveRaised',
            textStyle: TextStyle(fontSize: 40, color: Colors.green),
            depth: 10,
            style: ThreeDStyle.perspectiveRaised,
            angle: pi / 6,
            perspectiveDepth: 30,
          ),
          ThreeDText(
            text: 'perspectiveInset',
            textStyle: TextStyle(fontSize: 40, color: Colors.purpleAccent),
            depth: 40,
            style: ThreeDStyle.perspectiveInset,
            angle: pi / 6,
            perspectiveDepth: 20,
          ),
          ThreeDText(
            text: 'perspectiveRight',
            textStyle: TextStyle(fontSize: 25, color: Colors.yellow),
            style: ThreeDStyle.perspectiveLeft,
            perspectiveDepth: 45.0,
          ),
          ThreeDText(
            text: 'perspectiveLeft',
            textStyle: const TextStyle(
                fontSize: 25, color: Colors.pink, fontWeight: FontWeight.bold),
            depth: 6,
            style: ThreeDStyle.perspectiveLeft,
            perspectiveDepth: -45.0,
          ),
          ThreeDText(
            text: "inset",
            textStyle: TextStyle(
              fontSize: 64,
              color: Colors.pinkAccent,
            ),
            style: ThreeDStyle.inset,
          ),
          ThreeDText(
            text: 'raised',
            textStyle: TextStyle(fontSize: 64, color: Colors.blue),
            depth: 10,
            style: ThreeDStyle.raised,
          ),
          ThreeDText(
            text: 'standard',
            textStyle: TextStyle(fontSize: 40, color: Colors.blue),
            depth: 5,
            style: ThreeDStyle.standard,
          ),
          ThreeDText(
            text: 'inset with angle',
            textStyle: TextStyle(fontSize: 40, color: Colors.deepOrange),
            depth: 5,
            style: ThreeDStyle.inset,
            angle: 70,
            // 30 degrees
          ),
        ],
      )),
    );
  }
}
